src/machine/operation.c

```diff
@@ -21,60 +46,42 @@ static int operation_done(sd_event_source *s, const siginfo_t *si, void *userdat
 
         o->pid = 0;
 
-        if (si->si_code != CLD_EXITED) {
-                r = sd_bus_error_set(&error, SD_BUS_ERROR_FAILED, "Child died abnormally.");
-                goto fail;
-        }
+        if (o->message) {
+                _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
 
-        if (si->si_status == EXIT_SUCCESS)
-                r = 0;
-        else if (read(o->errno_fd, &r, sizeof(r)) != sizeof(r)) { /* Try to acquire error code for failed operation */
-                r = sd_bus_error_set(&error, SD_BUS_ERROR_FAILED, "Child failed.");
-                goto fail;
-        }
-
-        if (o->done) {
-                /* A completion routine is set for this operation, call it. */
-                r = o->done(o, r, &error);
+                r = operation_done_internal(si, o, &error);
                 if (r < 0) {
                         if (!sd_bus_error_is_set(&error))
                                 sd_bus_error_set_errno(&error, r);
 
-                        goto fail;
-                }
-
-        } else {
-                /* The default operation when done is to simply return an error on failure or an empty success
-                 * message on success. */
-                if (r < 0) {
-                        sd_bus_error_set_errno(&error, r);
-                        goto fail;
+                        r = sd_bus_reply_method_error(o->message, &error);
+                        if (r < 0)
+                                log_error_errno(r, "Failed to reply to dbus message: %m");
+                } else {
+                    r = sd_bus_reply_method_return(o->message, NULL);
+                    if (r < 0)
+                            log_error_errno(r, "Failed to reply to dbus message: %m");
                 }
-
-                r = sd_bus_reply_method_return(o->message, NULL);
+        } else if (o->link) {
+                r = operation_done_internal(si, o, /* userdata = */ NULL);
                 if (r < 0)
-                        log_error_errno(r, "Failed to reply to message: %m");
+                        (void) sd_varlink_error_errno(o->link, r);
+                else
+                        (void) sd_varlink_reply(o->link, NULL);
         }
 
         operation_free(o);
         return 0;
-
-fail:
-        r = sd_bus_reply_method_error(o->message, &error);
-        if (r < 0)
-                log_error_errno(r, "Failed to reply to message: %m");
-
-        operation_free(o);
-        return 0;
 }
 
-int operation_new(Manager *manager, Machine *machine, pid_t child, sd_bus_message *message, int errno_fd, Operation **ret) {
+int operation_new(Manager *manager, Machine *machine, pid_t child, sd_bus_message *message, sd_varlink *link, int errno_fd, Operation **ret) {
         Operation *o;
         int r;
 
         assert(manager);
         assert(child > 1);
-        assert(message);
+        assert(message || link);
+        assert(!(message && link));
```

Comment:
The two lines can be merged: `assert(!message != !link);`

Reply:
oooooooh, I have a tough time understanding this.  Maybe it is better to keep two lines where the logic is straightforward?!

Reply:
Sure. No problem.